### PR TITLE
Fixed issues with Constellix Provider

### PR DESCRIPTION
**Pull request by jrfrazier-tiggee**
All record types including A Record Pools can now be created.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -11,3 +11,6 @@ nosetests.xml
 octodns.egg-info/
 output/
 tmp/
+venv/
+.idea/
+.DS_Store
```
[the remark this answers is not in the document]
Comment (jrfrazier-tiggee) [reply .gitignore]: Thanks for pointing that out. I removed the 3 files from .gitignore

File: octodns/provider/constellix.py [hunk 2]
```diff
@@ -19,59 +19,108 @@
...
+    def current_time():
+        """
+        calculates the total milliseconds from epoch (UTC)
+
+        Returns:
+            string: current UTC time in milliseconds, as a string
+        """
+        return str(int(
+            (datetime.utcnow() - datetime(1970, 1, 1)).total_seconds() * 1000
+        ))
```
[the remark this answers is not in the document]
Comment (jrfrazier-tiggee) [reply octodns/provider/constellix.py]: We went with that as our system requires the current epoch time down to the millisecond. However, I fixed the code to utilize time() instead.

File: octodns/provider/constellix.py [hunk 3]
```diff
@@ -82,95 +128,226 @@ def _request(self, method, path, params=None, data=None):
...
+                else:
                     for v in value:
-                        v['value'] = self._absolutize_value(v['value'],
-                                                            zone_name)
+                        v['value'] = self._absolutize_value(
+                            v['value'], zone_name
+                        )
 
             # compress IPv6 addresses
-            if record['type'] == 'AAAA':
```
Comment (yzguy) [inline octodns/provider/constellix.py]: Probably still going to need this to keep compatibility between providers. Constellix seems to return the `0` between the colons in an IPv6 address when others keep it compressed

```
* constellix (ConstellixProvider)
*   Update
*     <AaaaRecord AAAA 3600, testyzguy.io., ['2001:470:abcd:53:0:0:0:2']> ->
*     <AaaaRecord AAAA 3600, test.yzguy.io., ['2001:470:abcd:53::2']> (config)
```

File: octodns/provider/constellix.py [hunk 4]
```diff
@@ -82,95 +128,226 @@ def _request(self, method, path, params=None, data=None):
...
+        :param zone_name: string - The domain name (required dot at end)
+        :return:
+            string: The absolute version of the FQDN
+        """
+        if value == '' or value == '.':
+            fqdn = zone_name
         elif not value.endswith('.'):
-            value = '{}.{}'.format(value, zone_name)
+            fqdn = '{}.{}'.format(value, zone_name)
+        else:
```
Comment (yzguy) [inline octodns/provider/constellix.py]: This is unnecessary and causes problems when the value is to something not within the same zone

```
* constellix (ConstellixProvider)
*   Update
*     <CnameRecord CNAME 3600, calendar.yzguy.io., ghs.googlehosted.com.yzguy.io.> ->
*     <CnameRecord CNAME 3600, calendar.yzguy.io., ghs.googlehosted.com.> (config)
```

File: octodns/provider/constellix.py [hunk 5]
```diff
@@ -191,51 +669,88 @@ def _data_for_multiple(self, _type, records):
...
+    @staticmethod
+    def _data_for_CAA(_type, records):
         values = []
         record = records[0]
         for value in record['value']:
             values.append({
                 'flags': value['flag'],
                 'tag': value['tag'],
-                'value': value['data']
+                # 'value': value['data']
```
Comment (yzguy) [inline octodns/provider/constellix.py]: This doesn't work, the commented out code is correct.

```
* constellix (ConstellixProvider)
*   Update
*     <CaaRecord CAA 3600, caa.yzguy.io., ['0 issue "{u'caaProviderId': 1, u'disableFlag': False, u'flag': 0, u'tag': u'issue', u'data': u'digicert.com'}"']> ->
*     <CaaRecord CAA 3600, caa.yzguy.io., ['0 issue "digicert.com"']> (config)
*   Summary: Creates=0, Updates=1, Deletes=0, Existing Records=48
```

File: octodns/provider/constellix.py [hunk 6]
```diff
@@ -19,59 +18,106 @@
...
-        return str(int(time.time() * 1000))
-
-    def _hmac_hash(self, now):
-        return hmac.new(self.secret_key.encode('utf-8'), now.encode('utf-8'),
-                        digestmod=hashlib.sha1).digest()
-
-    def _request(self, method, path, params=None, data=None):
-        now = self._current_time()
-        hmac_hash = self._hmac_hash(now)
+    @staticmethod
```
Comment (yzguy) [inline octodns/provider/constellix.py]: Not sure all these `@staticmethod` are necessary, most of the methods are only relevant within an instance of the ConstellixProvider class.

File: octodns/provider/constellix.py [hunk 7]
```diff
@@ -191,51 +669,88 @@ def _data_for_multiple(self, _type, records):
...
 
-    def _data_for_NS(self, _type, records):
+    @staticmethod
+    def _data_for_NS(_type, records):
         record = records[0]
         return {
             'ttl': record['ttl'],
             'type': _type,
-            'values': [value['value'] for value in record['value']]
+            'values': [r['value'] for r in records]
```
Comment (yzguy) [inline octodns/provider/constellix.py]: This causes an error, the previous code works as expected.

```
  File "/Users/yzguy/Projects/octodns/env/bin/octodns-sync", line 11, in <module>
    load_entry_point('octodns==0.9.8', 'console_scripts', 'octodns-sync')()
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/cmds/sync.py", line 39, in main
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/manager.py", line 301, in sync
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/manager.py", line 56, in result
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/manager.py", line 237, in _populate_and_plan
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/provider/base.py", line 49, in plan
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/provider/constellix.py", line 417, in populate
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/record/__init__.py", line 99, in new
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/record/__init__.py", line 261, in validate
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/record/__init__.py", line 976, in validate
AttributeError: 'list' object has no attribute 'endswith'
make: *** [noop] Error 1
```

File: octodns/provider/constellix.py [hunk 8]
```diff
@@ -258,8 +774,10 @@ def _data_for_single(self, _type, records):
         }
 
     _data_for_CNAME = _data_for_single
+    _data_for_PTR = _data_for_single
```
Comment (yzguy) [inline octodns/provider/constellix.py]: `_data_for_PTR`'s data structure is not the same as `_data_for_single`. The previous `_data_for_ALIAS` was correct

```
Traceback (most recent call last):
  File "/Users/yzguy/Projects/octodns/env/bin/octodns-sync", line 11, in <module>
    load_entry_point('octodns==0.9.8', 'console_scripts', 'octodns-sync')()
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/cmds/sync.py", line 39, in main
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/manager.py", line 301, in sync
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/manager.py", line 56, in result
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/manager.py", line 237, in _populate_and_plan
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/provider/base.py", line 49, in plan
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/provider/constellix.py", line 417, in populate
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/record/__init__.py", line 99, in new
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/record/__init__.py", line 360, in validate
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/record/__init__.py", line 710, in validate
AttributeError: 'list' object has no attribute 'endswith'
make: *** [noop] Error 1
```

File: octodns/provider/constellix.py [hunk 9]
```diff
@@ -328,16 +813,19 @@ def _params_for_multiple(self, record):
...
+    @staticmethod
+    def _params_for_single(record):
         yield {
             'name': record.name,
             'ttl': record.ttl,
             'host': record.value,
         }
 
     _params_for_CNAME = _params_for_single
+    _params_for_PTR = _params_for_single
```
Comment (yzguy) [inline octodns/provider/constellix.py]: `_params_for_PTR` are not the same as `_params_for_single`, the previous `_params_for_ALIAS` were correct

```
Traceback (most recent call last):
  File "/Users/yzguy/Projects/octodns/env/bin/octodns-sync", line 11, in <module>
    load_entry_point('octodns==0.9.8', 'console_scripts', 'octodns-sync')()
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/cmds/sync.py", line 39, in main
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/manager.py", line 330, in sync
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/provider/base.py", line 93, in apply
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/provider/constellix.py", line 651, in _apply
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/provider/constellix.py", line 589, in _apply_create
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/provider/constellix.py", line 246, in record_create
  File "build/bdist.macosx-10.14-x86_64/egg/octodns/provider/constellix.py", line 123, in _request
octodns.provider.constellix.ConstellixClientBadRequest:
  - Standard should have atleast 1 active value
```

File: octodns/provider/constellix.py [hunk 10]
```diff
@@ -347,23 +835,28 @@ def _params_for_ALIAS(self, record):
...
-    def _params_for_MX(self, record):
-        values = []
-        for value in record.values:
-            values.append({
-                'value': value.exchange,
-                'level': value.preference
-            })
         yield {
-            'value': value.exchange,
+            'value': values[-1]['value'],
```
Comment (yzguy) [inline octodns/provider/constellix.py]: This line is apparently not necessary and can be removed.

**Comment by yzguy** [comment]
I did a pass over the normal DNS record management, left some inline comments. I am unfamiliar with the pool stuff so I'll need a little more time to figure that out.

There are alot of `# pragma: no coverage` or `# pragma: no cover`, from my understanding this just says to exclude this line from test coverage. We would want tests for everything and not exclude things.

Additionally, we snuck in the fix for the ALIAS to ANAME in a previous PR, then there was some py3 fixes. You'll want to rebase onto master and fix up anything necessary.

**Pull request closed without merging**